modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java

```diff
@@ -1021,6 +1025,15 @@ private boolean isLocalNodeInBaseline() {
         return topology != null && topology.consistentIds().contains(cctx.localNode().consistentId());
     }
 
+    /**
+     * @return {@code true} if Event node is in baseline and {@code false} otherwise.
+     */
+    public boolean isEventNodeInBaseline() {
+        BaselineTopology top = cctx.discovery().discoCache().state().baselineTopology();
+
+        return top != null && top.consistentIds().contains(firstDiscoEvt.eventNode().consistentId());
+    }
```

Comment:
Looks like there is a race here. discovery().discoCache() will return a current snapshot, which is updated in discovery thread, while the method is executed in exchange thread. Returned disco cache may have topology version different from the exchange future topology version.

Reply:
BTW, what happens when exchanges are merged? Two failed nodes -> we need to check both? Failed and joined -> We should not allow fast-path exchange.

Reply:
1) Correct fix is to use versioned disco? 
`cctx.discovery().discoCache(topVer)...` correct?

2) We newer merge left with anything 
`if (baselineNodeLeft) baselineNodeLeft = true;
merge = false;`
so, even cascade failure will be processed one-by-one.

Reply:
1. Seems you can use `firstEvtDiscoCache`